### ISSUE TITLE
geos ios, relocatable .la files

### ISSUE THREAD

**Issue by ruslo** [ruslo/hunter#218]
Currently failing with error:

```
CMake Error at autotools-merge-lipo.cmake:234 (message):
  lipo execution failed: fatal error: lipo: can't figure out the architecture
  type of:
  /.../Build/multi-arch/arm64/bin/geos-config
```
##

**Comment by ruslo** [comment]
@pretyman About this particular issue: I think this script is not needed - Findgeos module doing exactly the same but in CMake-friendly fashion. I guess we can ignore it like `*.la` files.

In general: I think we can add `EXCLUDE` parameter here, e.g. `hunter_autotools_project(... EXCLUDE_FILES "bin/geos-config" ...)`.

**Comment by pretyman** [comment]
I don't think it is wise to temper with the build like this, removing files from it. I think we should make an effort to reproduce the full build inside `<root-id>/Install`.

The `geos-config` is in fact needed by `spatialite`, and won't compile without it, I have tried.

The `*.la` files in `Qt` are actually needed by the `odb` ORM framework library `odb-qt` at `./configure` execution time - this means it runs libtool during the ./configure execution. This is because the `libQt5Core.la` file has a field `dependency_libs=...` which contains its dependencies.

By ripping files out of the build, we increase the chance that projects which usually compiled together, will end up not compiling. This could bring frustration to users who want to integrate new packages: "I always compiled package `Y` with package `X` in my system, but Hunter can't compile the two together for some reason...".  

The more the build is untouched inside the `<root-id>/Install`, the higher the chances that other dependents projects will succeed building as well.

I think you have to re-evaluate this tactic of removing files from the build in the overall strategy of your project Hunter, based on who you want to target and how you want their interaction with your package manager should be. 

From the cache point of view, I re-iterate my suggestion that we could look into all text files in the resulting build searching the `@HUNTER_PACKAGE_INSTALL_DIR@`, keeping the files into a list which will be saved and replace the path with the placeholder `@HUNTER_PACKAGE_INSTALL_DIR@`. 

When hitting the cache, after uncompression is done, we look at the files in the saved list and replace the placeholder with its final destination. It would be nice to have a way of telling if a file is actually text, so we don't run the risk of breaking any binary files.

At the end, it is a design decision, and it is your call :-)

**Comment by ruslo** [comment]
> I think we should make an effort to reproduce the full build inside `<root-id>/Install`

I guess we have quite a few choices here

> From the cache point of view, I re-iterate my suggestion that we could look into all text files in the resulting build searching the @HUNTER_PACKAGE_INSTALL_DIR@

I still hope that we can avoid "all text files" rules by specifying the exact files we need per package (like `bin/geos-config`) or common pattern like `lib/*.la`. Also I guess it's not a bad strategy to even make `*.la` modification as accurate as possible (e.g. do update only `^libdir=` or `^dependency_libs=` patterns). Regarding `@HUNTER_PACKAGE_INSTALL_DIR@` variable I suppose you want to use `configure_file` command which also may introduce non-trivial bugs (what if file contain other `@...@` variables and they are not designed to be substituted?). An alternative will be using `string(REPLACE` command with special string. And since everything happens automatically we can choose very long and ugly name for such job, like: `__THIS_HUNTER_INTERNAL_VARIABLE_REPRESENTS_INSTALL_PREFIX__`

Applying this design to the current issue:
* mark `bin/geos-config` as a special file for patching
* patch every `multi-arch/${arch}/bin/geos-config` ( path -> __HUNTER_INTERNAL_...)
* test all `geos-config` are equals and install this script to `HUNTER_PACKAGE_INSTALL_PREFIX`
* patch back with new path ( __HUNTER_INTERNAL_... -> path) (note that package is not cacheable, for cacheable package we need additional logic)

**Comment by pretyman** [comment]
That is pretty good. Would you make any suggestions for a caching API to allow these features? I can help implement, I just don't think I can come up with something worthwhile for an API. If you define it, I can work on the implementation and then you can just brush it off. 

I believe that this should be in caching layer, because Qt could also benefit from it. But you might have others ideas.

**Comment by ruslo** [comment]
> Would you make any suggestions for a caching API to allow these features?

I guess this should work:
* Add extra argument to the `hunter_download` function, say `PACKAGE_UNRELOCATABLE_TEXT_FILES`
* After argument parsing variable name will be `HUNTER_PACKAGE_UNRELOCATABLE_TEXT_FILES`
* Variable can be used in `hunter_load_from_cache` function which is loaded only from `hunter_download`
* Patching code can be saved in some new function, say `hunter_expand_unrelocatable_text_files` (will be easy to test it) and must be done after unpack and before saving `DONE` stamp:
```cmake
hunter_unpack_directory("${archive_file}" "${HUNTER_INSTALL_PREFIX}")
hunter_expand_unrelocatable_text_files() # Use HUNTER_INSTALL_PREFIX and HUNTER_PACKAGE_UNRELOCATABLE_TEXT_FILES to make a `__HUNTER_INTERNAL_` -> `HUNTER_INSTALL_PREFIX` replace
file(WRITE "${HUNTER_PACKAGE_DONE_STAMP}" "")
return()
```

> But you might have others ideas.

The main idea is to solve it step by step (with testing and examples reproducing the bugs) so we can keep our brains non-damaged :) This issue is not touching cache and can be solved without new `hunter_expand_unrelocatable_text_files` module.

**Comment by ruslo** [comment]
To test: https://github.com/ruslo/hunter/commit/e3430ac72142d4f2914e7e0b885287ab54e50091

Usage example:
```cmake
hunter_pick_scheme(DEFAULT url_sha1_autotools)
hunter_cacheable(xorg-macros)
hunter_download(
    PACKAGE_NAME xorg-macros
    PACKAGE_UNRELOCATABLE_TEXT_FILES "share/pkgconfig/xorg-macros.pc"
)
```

**Comment by pretyman** [comment]
Thank you very much Ruslan, this feature was much in need!

I tried to start applying it to the Qt package, however Qt has different relocatable files when compiled with `Debug` configuration from the ones of the `Release`, for example:
```
lib/libQt5Core.la
lib/libQt5Core_debug.la
```

So I need to have a test somewhere to check `HUNTER_Qt_CONFIGURATION_TYPES`

I put in `cmake/projects/Qt/hunter.cmake`:

```CMake
if(HUNTER_Qt_CONFIGURATION_TYPES MATCHES "Release")
  set(qt_build_config_has_release TRUE)
endif()

if(HUNTER_Qt_CONFIGURATION_TYPES MATCHES "Debug")
  set(qt_build_config_has_debug TRUE)
endif()
```

Hoping to have access to the variables `qt_build_config_has_release` and `qt_build_config_has_debug` in Qt's modules such as qtbase - `cmake/projects/Qt/qtbase/hunter.cmake`, where I put:

```CMake
set(_qt_base_unrelocatable_files "")

if(UNIX)
  set(_la_files_that_can_have_debug_suffix
      Concurrent
      Core
      Gui
      Network
      OpenGL
      OpenGLExtensions
      PlatformSupport
      PrintSupport
      Sql
      Test
      Widgets
      Xml
  )

  list(APPEND _qt_base_unrelocatable_files "lib/libQt5Bootstrap.la")

  if(qt_build_config_has_release)
    foreach(_x ${_la_files_that_can_have_debug_suffix})
      list(APPEND _qt_base_unrelocatable_files
          "lib/libQt5${_x}.la"
      )
    endforeach()
  endif()

  if(qt_build_config_has_debug)
    foreach(_x ${_la_files_that_can_have_debug_suffix})
      list(APPEND _qt_base_unrelocatable_files
          "lib/libQt5${_x}_debug.la"
      )
    endforeach()
  endif()
endif()

hunter_pick_scheme(DEFAULT url_sha1_qt)
hunter_download(
    PACKAGE_NAME Qt
    PACKAGE_COMPONENT "qtbase"
    PACKAGE_INTERNAL_DEPS_ID "10"
    PACKAGE_UNRELOCATABLE_TEXT_FILES "${_qt_base_unrelocatable_files}"
)
```

However it didn't seem to pick up  `qt_build_config_has_release` and `qt_build_config_has_debug` as only `lib/libQt5Bootstrap.la` had changes applied to it.

Would you know how could I have this check of the `HUNTER_Qt_CONFIGURATION_TYPES` done only once and re-use its result in the different modules?

Thanks!

**Comment by ruslo** [comment]
> Hoping to have access to the variables qt_build_config_has_release and qt_build_config_has_debug in Qt's modules such as qtbase - cmake/projects/Qt/qtbase/hunter.cmake

Relying on the order of incuding of `hunter.cmake` files + interchanging of local variables is not a very good idea. What about new module for generating list of files?

Also note that all `hunter.cmake` modules except top one generated automatically from template `hunter.cmake.in`.

**Comment by ruslo** [comment]
> however Qt has different relocatable files when compiled with Debug configuration from the ones of the Release, for example

by the way does `*.la` relates to `*.pc`? Does it gives you any hints about #411?

**Comment by pretyman** [comment]
Somewhat related: https://github.com/ruslo/hunter/issues/207

**Comment by ruslo** [comment]
Patch and check [PACKAGE_UNRELOCATABLE_TEXT_FILES](https://github.com/ruslo/hunter/blob/2e44ccf581e8cff0e35372c598e95f491eac19cc/cmake/projects/libxml2/hunter.cmake#L38-L42):
- https://github.com/ruslo/hunter/blob/2e44ccf581e8cff0e35372c598e95f491eac19cc/scripts/autotools-merge-lipo.cmake.in#L44-L77

`bin/*`, `lib/*.a` and `lib/*.dylib` will be passed to `lipo`:
- https://github.com/ruslo/hunter/blob/2e44ccf581e8cff0e35372c598e95f491eac19cc/scripts/autotools-merge-lipo.cmake.in#L226-L239

The rest expected to be the same:
- https://github.com/ruslo/hunter/blob/2e44ccf581e8cff0e35372c598e95f491eac19cc/scripts/autotools-merge-lipo.cmake.in#L280-L304

`libxml2` example:
- multi-arch build: https://travis-ci.org/ingenue/hunter/jobs/524408225
- single-arch build: https://travis-ci.org/ingenue/hunter/jobs/524408226
